### PR TITLE
PR for BCIT Regression Code

### PR DESCRIPTION
**Pull request by youngk313**

### CODE REVIEW

File: watchmal/model/regression.py [hunk 1]
```diff
@@ -18,13 +18,14 @@ def forward(self, x):
         return x
 
 class LinearRegression(nn.Module):
-    def __init__(self, num_inputs, num_outputs):
+    def __init__(self, num_inputs, num_outputs, output):
```
Comment (aymuos15) [inline watchmal/model/regression.py:21]: Could you please elaborate on why you added the extra "output" parameter here? Thank you.

Comment (youngk313) [reply watchmal/model/regression.py:21]: This extra output parameter specifies what data to perform regression on (e.g. position or energy). The configurations for these output types are stored in 'model/regression_alg'. This is used in the regression engine to ensure the loss function compares to the correct output.

Comment (nickwp) [reply watchmal/model/regression.py:21]: Is there a reason why this parameter is part of the model class, rather than the engine? It seems it would only be used in the engine and not the model itself.

Comment (youngk313) [reply watchmal/model/regression.py:21]: Yes, you're right Nick, it is only used in the engine. I can make the change so that it feeds just into the engine.

Comment (nickwp) [reply watchmal/model/regression.py:21]: That would be good if you can, it would make sense for this to be a config parameter of the engine rather than the model I think. We hope to merge this PR quite soon and then continue working on regression, so if you don't have time to make these changes now we can always make another PR with that change in future.

Comment (youngk313) [reply watchmal/model/regression.py:21]: I can do that today.

Comment (nickwp) [reply watchmal/model/regression.py]: looks good, thanks for doing that!

Comment (aymuos15) [reply watchmal/model/regression.py]: Thank you for the clarifaction!

**Comment by nickwp** [comment]
I am merging this PR into the regression branch. Then we can continue working on it and create a new PR to merge into master when it's ready.

File: watchmal/engine/engine_regression.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+                loss = self.fit_transform(self.energies).to(self.device)
+            model_out = self.model(self.data)
+
+            self.loss = self.criterion(model_out, loss)
+
+        return {'loss': self.loss.detach().cpu().item(),
+                'output': model_out.detach().cpu().numpy(),
+                'raw_output': model_out}
+
+    def scale_positions(self, data):
```
Comment (aymuos15) [inline watchmal/engine/engine_regression.py:154]: - While re-running the file (as is) we faced an error of "invalid index" exactly after 1 whole epoch iterated.
-  Are these the same files used to report the findings in the final report? 
- We also got an error when we tried to train only with energies. 
- The scaling is used in the code as well (below). The scaling was not mentioned in any report. Is this something newly added or previously mentioned elsewhere?
@youngk313

Comment (aymuos15) [reply watchmal/engine/engine_regression.py:154]: Screenshot 2021-05-26 at 12 15 48 AM

Comment (youngk313) [reply watchmal/engine/engine_regression.py:154]: Hmm that is strange. I just ran it using positions today and I had no issues. As for the scaling, it was mentioned in the final report that is mentioned here: https://www.watchmal.org/projects/regression/bcit-meetings-2021. I will try running with energy and let you know if I encounter the same errors.

Comment (nickwp) [reply watchmal/engine/engine_regression.py:154]: Thanks for checking this. Can you confirm you're using exactly the version of the code in this pull request, and using the config that is the same as in this pull request?
